[PGO][GlobalValue][LTO]In GlobalValues::getGlobalIdentifier, use semicolon as delimiter for local-linkage varibles.

mingmingl-llvm opened the pull request:
Commit fe05193 (phab D156569), IRPGO names uses format `[<filepath>;]<linkage-name>` while prior format is `[<filepath>:<mangled-name>`. The format change would break the use case demonstrated in (updated)
`llvm/test/Transforms/PGOProfile/thinlto_indirect_call_promotion.ll` and `compiler-rt/test/profile/instrprof-thinlto-indirect-call-promotion.cpp`

This patch changes `GlobalValues::getGlobalIdentifer` to use the semicolon.

To elaborate on the scenario how things break without this PR
1. IRPGO raw profiles stores (compressed) IRPGO names of functions in one section, and per-function profile data in another section. The [NameRef](https://github.com/llvm/llvm-project/blob/fc715e4cd942612a091097339841733757b53824/compiler-rt/include/profile/InstrProfData.inc#L72) field in per-function profile data is the MD5 hash of IRPGO names.
2. When raw profiles are converted to indexed format profiles, the profiled address is [mapped](https://github.com/llvm/llvm-project/blob/fc715e4cd942612a091097339841733757b53824/llvm/lib/ProfileData/InstrProf.cpp#L876-L885) to the MD5 hash of the callee.
3. In `pgo-instr-use` thin-lto prelink pipeline, MD5 hash of IRPGO names will be [annotated](https://github.com/llvm/llvm-project/blob/fc715e4cd942612a091097339841733757b53824/llvm/lib/Transforms/Instrumentation/PGOInstrumentation.cpp#L1707) as value profiles, and used to import indirect-call-prom candidates. If the annotated MD5 hash is computed from the new format while import uses the prior format, the callee cannot be imported.

* `compiler-rt/test/profile/instrprof-thinlto-indirect-call-promotion.cpp` is added to have an end-to-end test.
* `llvm/test/Transforms/PGOProfile/thinlto_indirect_call_promotion.ll` is updated to have better test coverage from another aspect (as runtime tests are more sensitive to the environment and may be skipped by some contributors)

mingmingl-llvm committed:
[PGO][GlobalValue][LTO]In GlobalValues::getGlobalIdentifier, use semicolon as delimiter for local-linkage varibles.

Commit fe05193 (phab D156569), IRPGO names uses format
'[<filepath>;]<linkage-name>' while prior format is
[<filepath>:<linkage-name>'. The format change would break the use caes
demonstrated in (updated)
llvm/test/Transforms/PGOProfile/thinlto_indirect_call_promotion.ll

This patch changes GlobalValues::getGlobalIdentifer to use the
semicolon.

To elaborate on the scenario how things break without this PR
1. IRPGO raw profiles stores (compressed) IRPGO names of functions in
   one section, and per-function profile data in another section. One
   field in per-function profile data is the MD5 hash of IRPGO names.
2. When raw profiles are converted to indexed format profiles, the
   profiled address is mapped to the MD5 hash of the callee.
3. In thin-lto prelink pipeline, MD5 hash of IRPGO names will be
   annotated as value profiles, and used to import indirect-call-prom
   candidates. If the annotated MD5 hash is computed from the new format
   while import uses the prior format, the callee cannot be imported.

The updated test case Transforms/PGOProfile/thinlto_indirect_call_promotion.ll exercise the following path
- Annotate raw profiles and generate import summaries. Using the
  imported summaries, it tests that functions are correctly imported and
  ICP transformations happened.

teresajohnson commented on llvm/lib/ProfileData/InstrProf.cpp:
```diff
@@ -246,11 +246,27 @@ std::string InstrProfError::message() const {
 
-std::string getPGOFuncName(StringRef RawFuncName,
-                           GlobalValue::LinkageTypes Linkage,
+std::string getPGOFuncName(StringRef Name, GlobalValue::LinkageTypes Linkage,
```
If we're going to keep the original name, probably want to at least leave big comments in the header not to use it as it is legacy.

david-xl replied:
Also change the function name to getLegacyPGOFuncName

minglotus-6 replied:
I slightly prefer to use `getPGOFuncName` for `;` format and use `getLegacyPGOFuncName` for `:` format, and chose not to rename (want an NFC patch if rename is the way to go, since there are many existing callers).

I updated header and cpp, and did a minimal rename (this function has 3 callers, one in clang directory and another two in InstrProf.cpp)

teresajohnson replied:
Is this resolved? It looks like the most recent version of the patch changes getPGOFuncName to getLegacyPGOFuncName for the old format, and uses getIRPGOFuncName for the new format.

minglotus-6 replied:
>  It looks like the most recent version of the patch changes getPGOFuncName to getLegacyPGOFuncName for the old format, and uses getIRPGOFuncName for the new format.

This is intentional.  In this PR, renaming the `getPGOFuncName` (that takes [stringified names](https://github.com/llvm/llvm-project/blob/bc4e0c048aa3cd940b0cea787014c7e8680e5040/llvm/lib/ProfileData/InstrProf.cpp#L249) as argument) to  `getLegacyPGOFuncName` is probably fine since it has three callers.

Prior to this PR both `getPGOFuncName` (that [takes Functions as argument](https://github.com/llvm/llvm-project/blob/bc4e0c048aa3cd940b0cea787014c7e8680e5040/llvm/lib/ProfileData/InstrProf.cpp#L359)) and `getIRPGOFuncName` are called in quite a few places, so changing both names would cause too many diffs in this functional change. To use (simpler) `getPGOFuncName` for new `;` format, I'm thinking a separate NFC patch is the way to go.

teresajohnson replied:
Oh I missed the fact that one getPGOFuncName interface was left. I am only seeing 2 invocations of that interface outside of the unittest test. I would be in favor of doing renames of both getPGOFuncName interfaces in one patch. A separate NFC patch is a fine option, and keeps this patch just about fixing the ICP breakage caused by the delimiter change.

minglotus-6 replied:
> Is that still the case?

Clang FE-PGO calls this function, but afaik it's used for coverage rather than performance nowadays. I guess if Clang FE-PGO continues to use this function it shouldn't be called `legacy` in its name.

I'm wondering if it makes more sense if I rename `getPGOFuncName` (taking stringified names) to `getClangPGOFuncName` in this PR? Thoughts?

minglotus-6 replied:
undo the rename.

I didn't find a way to configure the Github notifications for real-time comments. Guess I'll need to combine email notifications.

teresajohnson commented on llvm/lib/ProfileData/InstrProf.cpp:
```diff
@@ -300,12 +316,8 @@ getIRPGONameForGlobalObject(const GlobalObject &GO,
-    Name.append(FileName.empty() ? "<unknown>" : FileName);
-    Name.append(";");
-  }
   Mangler().getNameWithPrefix(Name, &GO, /*CannotUsePrivateLabel=*/true);
```
Can we move the Mangler() handling into getGlobalIdentifier(), then remove the handling there of "Name[0] == '\1'" as I believe it is handled by Manger().getNameWithPrefix().

teresajohnson replied:
@ellishg How much of D156569 relied on the invocation of Mangler? It is not mentioned in the patch description, only the rationale for changing ":" to ";". The problem is if these are out of sync, then cross-module importing of indirectly called local functions will continue to be broken in whatever cases Mangler().getNameWithPrefix affects.

ellishg replied:
D156569 changes the format from `[<filepath>:]<function-name>` to `[<filepath>;]<linkage-name>`. Note the change from **function-name** to **linkage-name**. Having the mangled name is required so that we can pass a symbol order via `-symbol-ordering-file` or `-order_file`.

ellishg replied:
Well, I guess the name is still mangled, but I used `Mangler().getNameWithPrefix()` to get the linkage name.

minglotus-6 replied:
> I guess the name is still mangled, but I used Mangler().getNameWithPrefix() to get the linkage name.

When raw profiles stores (optionally filename prefixed) linkage names while global identifier uses the (filename prefixed) mangled name, the particular use case would be broken. Is it possible to convert objective-c mangled name to linkage name (needed by linker) in the post-processing of raw profiles? And do you think it make sense for the use case of D156569?

Say raw profiles still stores `[filepath;]mangled-name`, and `llvm-profdata` takes it to generate the linkage name that `-symbol-ordering-file` wants.

ellishg replied:
I think we can only generate the linkage name at compile time because we need to know the linkage, which `llvm-profdata` doesn't have.

teresajohnson replied:
Can you give an example of where this is called such that there is a difference between the function name and linkage name (at least for c++ these are the same afaik), and what that difference looks like? Are there specific callsites used for the symbol ordering support that need the mangling?

ellishg replied:
Here is the unittest for these functions which shows some examples. Even for c++ names there are differences

https://github.com/llvm/llvm-project/blob/70187ebadf20f014a7821cf322eae60192dbe4cc/llvm/unittests/ProfileData/InstrProfTest.cpp#L542-L609

ellishg replied:
> Was PGO broken in general before D156569? Or is this specific to -symbol-ordering-file and -order_file?

It wasn't broken in general, but it's needed to get `-order_file` working correctly.

> Also, it looks like some of the main changes here by the mangler for objective C are to strip the "\01" prefix. Is this different than the support to remove the '\1' mangling escape in getGlobalIdentifier()?

I think both remove the `\01` prefix. But the mangler might also prepend `_`/`l_` depending on the linkage, and `getGlobalIdentifier()` does not do this.

minglotus-6 replied:
>I think we can only generate the linkage name at compile time because we need to know the linkage, which llvm-profdata doesn't have.

Yep, but my original question is more about why using linkage-name in IRPGOFuncName (early in the pipeline) rather than storing it at a later phase of the pipeline. I read the code and [rfc](https://discourse.llvm.org/t/rfc-temporal-profiling-extension-for-irpgo/68068) again to make sense of why 'linkage-name' is needed in IRPGOFunc names.

> Was PGO broken in general before D156569? Or is this specific to -symbol-ordering-file and -order_file?
> It wasn't broken in general, but it's needed to get -order_file working correctly.

My understanding is that the specific data flow requires linkage-name to be stored in raw profiles. `llvm-profdata order` takes a profile and generates an order file (https://llvm.org/docs/CommandGuide/llvm-profdata.html#order); this file is used by `-symbol-ordering-file` directly.  Correct me if I miss anything.

A similar ELF option to `symbol-ordering-file` is [call-graph-ordering](https://github.com/llvm/llvm-project/blob/005c83380a907becbf5a6b4522fc43652c9536cd/lld/ELF/Driver.cpp#L1675-L1676), but with a different data flow. At a very high level, the edge (<caller, callee, weight> is passed in ELF sections, where linkage-name is already used.

teresajohnson replied:
> It wasn't broken in general, but it's needed to get -order_file working correctly.

Unfortunately this change broke aspects of ThinLTO ICP, however. Is it possible to change the handling of -order_file in the linker to modify the symbol names appropriately?

> To avoid subtle issues when linkage-name is different from mangled names,,I'm wondering if it warrants a change to use linkage-names (as opposed to mangled name) in GlobalValue::getGlobalIdentifier in this PR.

If the -order_name handling cannot be fixed as I suggested above, then yes, we need some solution to ensure that the hashed symbol names are consistent across PGO and ThinLTO.

ellishg replied:
I think it makes more sense to use linkage-names for IRPGO, `-order_file`, and ThinLTO. `-order_file` is used in the linker when it only knows linkage-names, so I don't think it makes sense to feed it mangled names.

minglotus-6 replied:
>  I think it makes more sense to use linkage-names for IRPGO, -order_file, and ThinLTO. -order_file is used in the linker when it only knows linkage-names, so I don't think it makes sense to feed it mangled names.

Thanks for the input.  (not to bikeshed but) I think for the purpose of computing global identifier, unique names should suffice. Clang-FE mangled names are unique so sounds fine.

Using linkage-name would be a non-trivial change, given the static `getGlobalIdentifier` takes a stringified name currently, and using linkage-name means requiring compatible change in each callsite (e.g., if the caller context doesn't have `GlobalValue` but just stringified names in the bitcode, make sure linkage-name exists in the bitcode, [this](https://github.com/llvm/llvm-project/blob/d6fbd96e5eaf3e8acbf1b43dce7a311352907567/llvm/lib/Bitcode/Reader/BitcodeReader.cpp#L6763) seems one example). An alternative fixup is to store the MD5 of `[filename;]linkage-name` (this is what's currently stored in this [field](https://github.com/llvm/llvm-project/blob/0f45e45847a5f2969b8021c787a566531fc96473/compiler-rt/include/profile/InstrProfData.inc#L72-L74)) and the MD5 of `[filename:]mangled-name` (the original hash before D156569) in the raw profiles, so `llvm-profdata` could choose properly (former for ordering and latter for ICP)

Would you mind if I create a Github issue to track how to fix other potential cases and assign it to you ? This PR would solve the colon and semicolon difference.

minglotus-6 replied:
Thanks for giving the thumb-up Ellis! Just created https://github.com/llvm/llvm-project/issues/74565.

teresajohnson commented on llvm/test/Transforms/PGOProfile/Inputs/update_icall_promotion_inputs.sh:
```diff
@@ -0,0 +1,70 @@
+EOF
+
+# Creates lib.cc. global_func might call one of two indirect callees. Both
+# indirect callees have internal linkage.
```
Are we now missing testing of ThinLTO ICP importing for a global function? What if one of the functions was left with non-local linkage like before?

teresajohnson reviewed (commented):
Thanks! A few comments below.

mingmingl-llvm committed:
address feedback

mingmingl-llvm committed:
keep getPGOFuncName and document it as front-end instrumentation only

minglotus-6 commented:
Using the same added ICP test,  profile matching on local-linkage `_ZL7callee0v` using `clang++ -v -fuse-ld=lld -O2 -fprofile-use=thinlto_icall_prom.profdata <file-list>` , as [this](https://gist.github.com/minglotus-6/11817ba645c6b12cd7116f41bfb1185e) pgo-instr-use output shows.

However, when trying to follow the counter matching [code path](https://github.com/llvm/llvm-project/blob/df7545e4be5cb65ef3ddd278119c0b8b2f37b0ec/llvm/lib/Transforms/Instrumentation/PGOInstrumentation.cpp#L1337), I came across another place(https://github.com/llvm/llvm-project/blob/df7545e4be5cb65ef3ddd278119c0b8b2f37b0ec/llvm/lib/ProfileData/InstrProfReader.cpp#L1019) that might have something to do with colon/semicolon delimiter inside itanium remapper (https://github.com/llvm/llvm-project/blob/df7545e4be5cb65ef3ddd278119c0b8b2f37b0ec/llvm/lib/ProfileData/InstrProfReader.cpp#L1055). I'm wondering if this needs an update (with updated regression test)? @david-xl @xur-llvm who might have more context.

teresajohnson commented:
> Using the same added ICP test, profile matching on local-linkage `_ZL7callee0v` using `clang++ -v -fuse-ld=lld -O2 -fprofile-use=thinlto_icall_prom.profdata <file-list>` , as [this](https://gist.github.com/minglotus-6/11817ba645c6b12cd7116f41bfb1185e) pgo-instr-use output shows.

Can you clarify what you are saying here - is it working or not working?

minglotus-6 commented:
> Can you clarify what you are saying here - is it working or not working?

It's working. The local-linkage `_ZL7callee0v` has function entry count annotated as `!prof`.

David says the itanium remapper file was only used once during gcc to llvm transition, so not relevant here.

minglotus-6 commented:
Just noticed there is a merge conflict. Will update my fork and merge.

mingmingl-llvm committed:
git merge main and resolve merge conflict

teresajohnson commented:
> David says the itanium remapper file was only used once during gcc to llvm transition, so not relevant here.

I believe it was actually for the libstdc++ to libc++ transition (see https://reviews.llvm.org/D51247 and https://reviews.llvm.org/D51240).

If it is broken we'll at least want to add a FIXME there.

david-xl commented:
> > David says the itanium remapper file was only used once during gcc to llvm transition, so not relevant here.
> 
> I believe it was actually for the libstdc++ to libc++ transition (see https://reviews.llvm.org/D51247 and https://reviews.llvm.org/D51240).
> 
> If it is broken we'll at least want to add a FIXME there.

Yes, I meant libstdc++ to libc++ transition.  Why source line is this comment addressing? I take take a look the changes/comments there.

minglotus-6 commented:
> > > David says the itanium remapper file was only used once during gcc to llvm transition, so not relevant here.
> > 
> > 
> > I believe it was actually for the libstdc++ to libc++ transition (see https://reviews.llvm.org/D51247 and https://reviews.llvm.org/D51240).
> > If it is broken we'll at least want to add a FIXME there.
> 
> Yes, I meant libstdc++ to libc++ transition. Why source line is this comment addressing? I take take a look the changes/comments there.

Sorry for the misinformation. I think the itanium remapper needs a `:` -> `;` update (going to update this PR and tests to do it), since (for local-linkage functions) the function name used to look up profiles should use `;` delimiter.

mingmingl-llvm committed:
Use semicolon in itanium symbol remapper

mingmingl-llvm committed:
refine code comments and use semicolon in ItaniumRemapper's handling

minglotus-6 commented:
> > > > > David says the itanium remapper file was only used once during gcc to llvm transition, so not relevant here.
> > > > 
> > > > 
> > > > I believe it was actually for the libstdc++ to libc++ transition (see https://reviews.llvm.org/D51247 and https://reviews.llvm.org/D51240).
> > > > If it is broken we'll at least want to add a FIXME there.
> > > 
> > > 
> > > Yes, I meant libstdc++ to libc++ transition. Why source line is this comment addressing? I take take a look the changes/comments there.
> > 
> > 
> > Sorry for the misinformation, and thanks for the Phab links.
> > I think the itanium remapper needs a `:` -> `;` update (going to update this PR and related tests), since (for local-linkage functions) the function name used to look up profiles should use `;` delimiter.
> 
> > > > > David says the itanium remapper file was only used once during gcc to llvm transition, so not relevant here.
> > > > 
> > > > 
> > > > I believe it was actually for the libstdc++ to libc++ transition (see https://reviews.llvm.org/D51247 and https://reviews.llvm.org/D51240).
> > > > If it is broken we'll at least want to add a FIXME there.
> > > 
> > > 
> > > Yes, I meant libstdc++ to libc++ transition. Why source line is this comment addressing? I take take a look the changes/comments there.
> > 
> > 
> > Sorry for the misinformation, and thanks for the Phab links.
> > I think the itanium remapper needs a `:` -> `;` update (going to update this PR and related tests), since (for local-linkage functions) the function name used to look up profiles should use `;` delimiter.
> 
> The remapper is not aware of any internal symbol mangling scheme, so those entires won't be tracked by it. In other words, there is no need to change anything there, I think.

Not updating Itanium remapper should work for PGO counter matching until the next transition (details below); for consistency I just updated Itanium remapper's `extractName` to use semicolon.

The details
* For PGO counter matching, instr prof reader [asks](https://github.com/llvm/llvm-project/blob/32ec5fbfed32f37aa070ee38e9b038bd84ca6479/llvm/lib/ProfileData/InstrProfReader.cpp#L1339) read remapper for a record. 
* Only with a remapping file provided (specified by `-fprofile-remapping-file`), a itanium remap reader is constructed. And when remapping file is not specified, a no-op remap reader is constructed. [source code](https://github.com/llvm/llvm-project/blob/32ec5fbfed32f37aa070ee38e9b038bd84ca6479/llvm/lib/ProfileData/InstrProfReader.cpp#L1306-L1314) 
* When remapping file is specified, remap reader tries to extract the mangled name (removing `filename` prefix`) by finding a `:` (no longer used as delimiter for newer profiles) and remaps the mangled name. If `:` is not updated to `;`, the name is remapped to itself (irpgo func format) and profiles could still be found. However, not updating means remapping becomes no-op for local-linkage functions, which is fine after the transition complete but doesn't work for new transitions (if it happens..)

ellishg commented on llvm/test/Transforms/PGOProfile/Inputs/a.out:
Please remember to remove the binaries that are not used.

minglotus-6 replied:
done. Sorry about leaving debugging-only stuff in a commit.

ellishg commented on llvm/test/Transforms/PGOProfile/thinlto_indirect_call_promotion.ll:
```diff
@@ -1,39 +1,45 @@
-; Do setup work for all below tests: generate bitcode and combined index
-; RUN: opt -module-summary %s -o %t.bc
-; RUN: opt -module-summary %p/Inputs/thinlto_indirect_call_promotion.ll -o %t2.bc
+; The raw profiles and reduced IR inputs are generated from Inputs/update_icall_promotion_inputs.sh
```
Can we create a `.proftext` file and use it for this test instead of generating it from the script above?

teresajohnson replied:
No because the point of this test is to catch regressions if the separator character (which is hardcoded in the proftext) changes again in the future.

minglotus-6 replied:
Using a `.proftext` file (and convert it to indexed profiles in the test using `RUN` line) generally has better readability.

For better test coverage, I changed to store raw profile data in `Inputs/` directory and added some comments at the beginning of test case Transforms/PGOProfile/thinlto_indirect_call_promotion.ll

minglotus-6 replied:
The script and c++ source is provided for on-demand update (e.g., version bump, etc). IIUC the goal of a `compiler-rt` test is to generate raw profiles from C++ source for each run of the IR test (`Transforms/PGOProfile/thinlto_indirect_call_promotion.ll`) (i.e.,the IR test would use the the output from the `compiler-rt` test). 

Practically, is there a way to run two tests one after the other so one could make use of the output of the other? Or shall the `compiler-rt` test verify the content of (raw or indexed) profile content?   (I'm assuming the IR test itself stays in `llvm/test/Transforms/PGOProfile` directory makes more sense here)

teresajohnson replied:
The reason for having this in a single test this way is specifically to detect mismatches in the separator used in the PGO name and in the ThinLTO index. For this we need to convert from raw profile and feed it through ThinLTO+ICP. 

Having separate tests would not do this. I.e. with the change from ':' to ';', the compiler-rt test would presumably have failed and would be updated, but the test using the proftext input would not fail. The lack of testing these pieces together led to the prior separator rename not updating all the necessary compiler bits.

ellishg commented on llvm/test/Transforms/PGOProfile/thinlto_indirect_call_promotion.ll:
```diff
@@ -1,39 +1,45 @@
+; Do setup work for all below tests: annotate value profiles, generate bitcode and combined index.
+; Explicitly turn off ICP pass in Inputs/thinlto_indirect_call_promotion.ll.
+; This way ICP happens in %t.bc after _Z11global_funcv and two indirect callees are imported.
+; RUN: opt -passes=pgo-instr-use -pgo-test-profile-file=%p/Inputs/thinlto_icall_prom.profdata -module-summary %s -o %t.bc
```
I think it's a bit cleaner to use [`split-file`](https://llvm.org/docs/TestingGuide.html#extra-files) and have all the `.ll` code in this file instead of putting different files in `Inputs/`. You can even have the `.proftext` file here if that works.

minglotus-6 replied:
Thanks for the pointer to split-file. Applied it for reduced IR. Choose to use raw profiles (see the other comment) per discussion.

MaskRay replied:
A common pattern is `rm -rf %t && split-file %s %t`. When dealing with multiple files, it is useful to ensure the temporary directory is cleared.  If specifying `%t/out.bc` is too inconvenient, you can run `rm -rf %t && split-file %s %t && cd %t` and then forget about `%t`.

mingmingl-llvm committed:
resolve feedback

minglotus-6 commented:
> . For IR PGO, there is basically no need to do so as the instrumentation and profile-use should be in-sync. For front-end instrumentation, there seem to be some use cases to use out of sync profile: https://reviews.llvm.org/D51240.

Thanks for double checking. I noticed the ICP and stale profile tolerance discussions when read the Phab history; it's good Phab review history are still available nowadays.

IRPGO profiles could be used along with supplementary sample-pgo profiles. I'll probably read relevant code in llvm-profdata to understand how these interact in theory mostly for my own curiosity (hopefully no rough edges as long as  `llvm-profdata` uses the same pgo name format used by latest compiler)

minglotus-6 commented:
> > . For IR PGO, there is basically no need to do so as the instrumentation and profile-use should be in-sync. For front-end instrumentation, there seem to be some use cases to use out of sync profile: https://reviews.llvm.org/D51240.
> 
> Thanks for double checking. I noticed the ICP and stale profile tolerance discussions when read the Phab history; it's good Phab review history are still available nowadays.
> 
> IRPGO profiles could be used along with supplementary sample-pgo profiles. I'll probably read relevant code in llvm-profdata to understand how these interact in theory mostly for my own curiosity (hopefully no rough edges as long as `llvm-profdata` uses the same pgo name format used by latest compiler)

For irpgo with supplementary profiles, this line to build a map (https://github.com/llvm/llvm-project/blob/44dc1e0baae7c4b8a02ba06dcf396d3d452aa873/llvm/tools/llvm-profdata/llvm-profdata.cpp#L982) needs an update. Will do it together with the test [update](https://github.com/llvm/llvm-project/pull/74008#discussion_r1421018997) in this pull request.

david-xl commented:
> > > . For IR PGO, there is basically no need to do so as the instrumentation and profile-use should be in-sync. For front-end instrumentation, there seem to be some use cases to use out of sync profile: https://reviews.llvm.org/D51240.
> > 
> > 
> > Thanks for double checking. I noticed the ICP and stale profile tolerance discussions when read the Phab history; it's good Phab review history are still available nowadays.
> > IRPGO profiles could be used along with supplementary sample-pgo profiles. I'll probably read relevant code in llvm-profdata to understand how these interact in theory mostly for my own curiosity (hopefully no rough edges as long as `llvm-profdata` uses the same pgo name format used by latest compiler)
> 
> For irpgo with supplementary profiles, this line to build a map (
> 
> https://github.com/llvm/llvm-project/blob/44dc1e0baae7c4b8a02ba06dcf396d3d452aa873/llvm/tools/llvm-profdata/llvm-profdata.cpp#L982
> 
> ) needs an update. Will do it together with the test [update](https://github.com/llvm/llvm-project/pull/74008#discussion_r1421018997) in this pull request.

A follow up patch to fix the tool is fine too (or may be better).

MaskRay commented on llvm/lib/IR/Globals.cpp:
```diff
@@ -144,25 +145,27 @@ void GlobalObject::copyAttributesFrom(const GlobalObject *Src) {
-      NewName = NewName.insert(0, FileName.str() + ":");
+      GlobalName.append(FileName.str());
+
+    GlobalName.append({kGlobalIdentifierDelimiter});
```
Nit; use `+=` or `push_back`

MaskRay commented on llvm/test/Transforms/PGOProfile/Inputs/update_thinlto_indirect_call_promotion_inputs.sh:
```diff
@@ -0,0 +1,60 @@
+fi
+
+# Allows the script to be invoked from other directories.
+OUTDIR=$(dirname $(realpath -s $0))
```
Nice script!
The clean up commands may be simplified using this scheme

```
#!/bin/bash
OUTDIR=$(dirname $(realpath -s $0))

dir=$(mktemp -d)
trap "rm -r $dir" EXIT
cd $dir
cat > lib.h <<EOF
...
EOF
...
rm -f $OUTDIR/thinlto_indirect_call_promotion.profraw
LLVM_PROFILE_FILE=$OUTDIR/thinlto_indirect_call_promotion ./a.out
```

MaskRay commented:
Trying to summarize the issue.

`__llvm_prf_nm` contains an entry `getIRPGONameForGlobalObject() == "lib.cc;_ZL7callee0v"`. The hash in the "VP" metadata uses `md5("lib.cc;_ZL7callee0v")`: `!31 = !{!"VP", i32 0, i64 1, i64 8947761083887884635, i64 1}`

Without this patch, when invoking `opt -passes=pgo-instr-use -pgo-test-profile-file=%t.profdata -module-summary %tdir/lib.ll -o %t2.bc`,
in `llvm/lib/Analysis/ModuleSummaryAnalysis.cpp:computeFunctionSummary`,
the local linkage `_ZL7callee0v`'s GUID is incorrectly computed as `md5("lib.cc:_ZL7callee0v")` and therefore is not associated with the global variable summary `_Z7callee1v`
(displayed as a stray `^2 = gv: (guid: 8947761083887884635)`).

This patch fixes `_ZL7callee0v`'s GUID to be `md5("lib.cc;_ZL7callee0v")`, correctly associating it with the global variable summary `_Z7callee1v`.

---

Mach-O (and 32-bit Windows) mangle global symbols with a leading `_`.
`llvm/lib/Analysis/ModuleSummaryAnalysis.cpp:computeFunctionSummary` uses `getGlobalIdentifier`, which does not have `GlobalObject` information for the mangling scheme.
The identifier can be different from the producer, causing missed import of functions for indirect-call-promotion (#74565).

mingmingl-llvm committed:
resolve feedbacks
- Add thinlto indirect-call-promotion regression test as a compiler-rt
  test and remove the original IR test.
  - `clangxx_pgouse` is added in complier-rt lit config for test use.
- In lib/IR/Globals.cpp, use '+=' operator for previous 'append' calls.
- In llvm/ProfileData/InstrProf.h, update a few function comments. These
  comments used to refer to `getPGOFuncName` while the actual name could
  be returned by `getPGOFuncName` (clang instrumentation) or
  `getIRPGOFuncName` (ir instrumentation).

minglotus-6 commented:
> Mach-O (and 32-bit Windows) mangle global symbols with a leading `_`.

It's very helpful to know the name differences exist on Mach-O/COFF to construct a test case for issue 74565.

https://godbolt.org/z/686Y9vYod shows the name difference (cross comparing IR and machine assembly, with and without `--target arm64-apple-macosx`), meaning the compiler-rt test (to be added in this pull-request) could be a test case for issue 74565.

The summary captures key points of the bug. Just to clarify on one detail, the md5 hash is actually computed (or more accurately [mapped](https://github.com/llvm/llvm-project/blob/fc715e4cd942612a091097339841733757b53824/llvm/lib/ProfileData/InstrProf.cpp#L876-L885) from a profiled address to the [NameRef field](https://github.com/llvm/llvm-project/blob/fc715e4cd942612a091097339841733757b53824/compiler-rt/include/profile/InstrProfData.inc#L72) of  during `llvm-profdata merge <raw-profile> -o <indexed-profile>`) and [annotated](https://github.com/llvm/llvm-project/blob/fc715e4cd942612a091097339841733757b53824/llvm/lib/Transforms/Instrumentation/PGOInstrumentation.cpp#L1707) on an indirect-call instruction by `pgo-instr-use` pass. Module-summary-analysis pass makes use of this hash by [extracting](https://github.com/llvm/llvm-project/blob/fc715e4cd942612a091097339841733757b53824/llvm/lib/Analysis/ModuleSummaryAnalysis.cpp#L454-L456) it from value profile metadata.

mingmingl-llvm committed:
make clang-format happy

MaskRay commented on llvm/lib/IR/Globals.cpp:
```diff
@@ -144,25 +145,27 @@ void GlobalObject::copyAttributesFrom(const GlobalObject *Src) {
     Name = Name.substr(1);
 
-  std::string NewName = std::string(Name);
+  SmallString<64> GlobalName;
```
If Name is longer than 64, this would cause two head allocations. It might be better to stick with `std::string`.

MaskRay commented on compiler-rt/test/profile/instrprof-thinlto-indirect-call-promotion.cpp:
```diff
@@ -0,0 +1,97 @@
+// expected !PGOFuncName metadata and external function callee1 doesn't have
+// !PGOFuncName metadata. Explicitly skip ICP pass to test ICP happens as
+// expected in the IR module that imports functions from lib.
+// RUN: %clang -target x86_64-unknown-linux-gnu -mllvm -disable-icp -fprofile-use=main.profdata -flto=thin -O2 -c lib.cpp -o lib.bc
```
`-target ` is deprecated. Use `--target=`. Actually, runtime tests don't usually set a `--target=`

MaskRay replied:
To match 32-bit Windows triples, `i.86.*windows` suffices. The normalized triples don't use `win32`.

MaskRay commented on compiler-rt/test/profile/instrprof-thinlto-indirect-call-promotion.cpp:
```diff
@@ -0,0 +1,97 @@
+//
+// Do setup work for all below tests.
+// Generate raw profiles from real programs and convert it into indexed profiles.
+// RUN: %clangxx_pgogen -fuse-ld=lld -O2 lib.cpp main.cpp -o main
```
lld requires `// REQUIRES: lld-available`

minglotus-6 replied:
I ended up removing `-fuse-ld=lld` in this test.

MaskRay replied:
If the default linker is GNU ld or gold, LLVMgold.so is needed to use LTO. LLVMgold.so is disabled for some builds, so we still need a `REQUIRES`. I think it is more robust to just require `// REQUIRES: lld-available` and use `-fuse-ld=lld`.

minglotus-6 replied:
I see. This makes sense to me. Done.

MaskRay commented on compiler-rt/test/profile/instrprof-thinlto-indirect-call-promotion.cpp:
```diff
@@ -0,0 +1,97 @@
+// RUN: opt main.import.bc -icp-lto -passes=pgo-icall-prom -S -pass-remarks=pgo-icall-prom 2>&1 | FileCheck %s --check-prefixes=ICP-IR,ICP-REMARK --implicit-check-not="!VP"
+
+// IMPORTS: main.cpp: Import _Z7callee1v
+// IMPORTS: main.cpp: Import _ZL7callee0v.llvm.{{[0-9]+}}
```
`{{[0-9]+}}` can be simplified to `[[#]]`

MaskRay commented on compiler-rt/test/profile/instrprof-thinlto-indirect-call-promotion.cpp:
```diff
@@ -0,0 +1,97 @@
+// PGOName: define internal void @_ZL7callee0v() {{.*}} !prof !{{[0-9]+}} !PGOFuncName ![[MD:[0-9]+]] {
+// PGOName: ![[MD]] = !{!"{{.*}}lib.cpp;_ZL7callee0v"}
+
+// IR-LABEL: define available_externally {{.*}} void @_Z11global_funcv() {{.*}} !prof !35 {
```
`!35` in `!prof !35`  may make the test too sensitive.

MaskRay commented on compiler-rt/test/profile/lit.cfg.py:
```diff
@@ -147,6 +147,9 @@ def exclude_unsupported_files_for_aix(dirname):
     ("%clang_pgouse=", build_invocation(clang_cflags) + " -fprofile-use=")
 )
+config.substitutions.append(
+    ("%clangxx_pgouse=", build_invocation(clang_cxxflags) + " -fprofile-use=")
```
Is this used?

MaskRay replied:
`rg clangxx_pgouse compiler-rt/test/` => no result?

minglotus-6 replied:
oh you are correct.

I once used `clangxx_pgouse` in the new test for profile use, but later thought `clang -c` (run up to assembler phase to generate bitcode) together with `opt` and `llvm-lto` is better.

MaskRay commented on llvm/include/llvm/IR/GlobalValue.h:
```diff
@@ -41,6 +41,10 @@ namespace Intrinsic {
 } // end namespace Intrinsic
 
+// Choose ';' as the delimiter. ':' was used once but it doesn't work well for
+// Objective-C functions which commonly have :'s in their names
```
add a full stop

MaskRay commented on llvm/test/Transforms/PGOProfile/thinlto_indirect_call_promotion.ll:
```diff
@@ -1,39 +0,0 @@
-; Do setup work for all below tests: generate bitcode and combined index
```
Hmm, I think it is still preferred to have a test here as runtime tests are more sensitive to the environment and may be skipped by some contributors.

It means we will have to duplicate some file between here and the runtime test. Perhaps adding a comment can help connect the two tests.

minglotus-6 replied:
this is a valid point. Add the test back (in the new commit), use `rm -rf %t && split-file %s %t && cd %t` on IR and apply the `trap-on-exit` clean-up on the shell script.

MaskRay replied:
It looks like that we had 3 precanned profraw files and this patch is adding another one. Since the raw profile isn't stable, this is going to add some inconvenience whenever the version is bumped.
`thinlto_indirect_call_promotion.profraw` probably should be rewritten as `.proftext` and `llvm-profdata merge`
```
% fd -e profraw
Transforms/PGOProfile/Inputs/thinlto_indirect_call_promotion.profraw
tools/llvm-profdata/Inputs/basic.profraw
tools/llvm-profdata/Inputs/compressed.profraw
tools/llvm-profdata/Inputs/c-general.profraw
```

It seems that there is an earlier comment that proftext may not cover the use case well, so we also utilize a runtime test in `compiler-rt/test`.

teresajohnson replied:
Without use of the raw profile, this test would not have caught the regression. If we think the new compiler-rt test is enough to catch this case in the future, then perhaps we don't need this LLVM test at all (there should already be some ICP tests). But if the compiler-rt test is not enough, because some people don't run those, then this should stay and use a raw profile as input.

MaskRay reviewed (commented):
Mostly looks good with some nits...

mingmingl-llvm committed:
resolve feedbacks

mingmingl-llvm committed:
resolve feedback

minglotus-6 commented on compiler-rt/test/profile/instrprof-thinlto-indirect-call-promotion.cpp:
```diff
@@ -0,0 +1,115 @@
+
+// Test should fail where linkage-name and mangled-name diverges, see issue https://github.com/llvm/llvm-project/issues/74565).
+// Currently, this name divergence happens on Mach-O object file format, or on
+// many (but not all) 32-bit Windows systems.
```
Some more details regarding this statement

1. The global prefix is `_` for {`MachO`, `WinCOFFX86`} and  `\0` (i.e., no prefix) for the rest of mangling modes ([source code](https://github.com/llvm/llvm-project/blob/185302530847a28f44e48a67a79fd4eba048a1c7/llvm/include/llvm/IR/DataLayout.h#L316-L328)). 
    * The global prefix is [used by IR Mangler](https://github.com/llvm/llvm-project/blob/62b21c6ced918c7fec97b557e3087e3ffdf71494/llvm/lib/IR/Mangler.cpp#L144) in global symbols (e.g., prefix [emitted](https://github.com/llvm/llvm-project/blob/62b21c6ced918c7fec97b557e3087e3ffdf71494/llvm/lib/IR/Mangler.cpp#L55-L56) if it's not `\0`)
2.  In the data-layout string,`m:o` specifies `MachO`, `m:x` specifies `WinCOFFX86`, and `m:e` specifies `ELF` ([parser source code](https://github.com/llvm/llvm-project/blob/185302530847a28f44e48a67a79fd4eba048a1c7/llvm/lib/IR/DataLayout.cpp#L510-L541)) 

With 1 and 2, note not all 32-bit windows uses `WinCOFFX86` mangling-mode. For instance, one windows-32 target [specifies](https://github.com/llvm/llvm-project/blob/a8ef9c0969ab0807672bcad60970bf22395ffaf1/clang/lib/Basic/Targets/X86.h#L672) `m:e`, and some [1] windows-32 target chooses between `m:e` and one of {`m:x`, `m:o`}

[1] [example 1](https://github.com/llvm/llvm-project/blob/a8ef9c0969ab0807672bcad60970bf22395ffaf1/clang/lib/Basic/Targets/X86.h#L581) chooses between `m:x` and `m:e`,and [example 2](https://github.com/llvm/llvm-project/blob/a8ef9c0969ab0807672bcad60970bf22395ffaf1/clang/lib/Basic/Targets/X86.h#L444-L447) chooses between `m:o` and `m:e`

mingmingl-llvm committed:
add .* suffix for 32-bit windows

MaskRay reviewed (approved):
LGTM once these changes are made.

* unused `%clangxx_pgouse=`
* profraw => proftext for the PGOProfile test
* lld-available -fuse-ld=lld and windows triple simplification for the compiler-rt test

It seems useful to wait for others' opinions as well.

teresajohnson commented on compiler-rt/test/profile/instrprof-thinlto-indirect-call-promotion.cpp:
```diff
@@ -0,0 +1,115 @@
+// IR-NEXT:  %1 = load ptr, ptr getelementptr inbounds ([2 x ptr], ptr @calleeAddrs,
+// IR-NEXT:  tail call void %1(), !prof ![[#PROF2:]]
+
+// The GUID of indirect callee is the MD5 hash of `/path/to/lib.cpp:_ZL7callee0v`
```
should this have the semicolon separator not a colon?

teresajohnson commented on llvm/lib/ProfileData/InstrProf.cpp:
```diff
@@ -352,6 +366,8 @@ std::string getIRPGOFuncName(const Function &F, bool InLTO) {
   return getIRPGOObjectName(F, InLTO, getPGOFuncNameMetadata(F));
 }
 
+// DEPRECATED. Use `getIRPGOFuncName`for new code. See that function for
```
In the header it is described as for FE instrumentation. Probably want the comments to be consistent?

minglotus-6 replied:
done.

teresajohnson reviewed (approved):
LGTM other than a couple of minor comments and pending resolution of the LLVM IR test. Thanks!

mingmingl-llvm committed:
resolve feedbacks

mingmingl-llvm reviewed (commented):
Resolved comments except the pending discussion about whether to use `.proftext` or`.profraw`.

Added `REQUIRES: zlib` for LLVM IR test since the profile reader should be built with zlib support. 

I'll probably spend sometime to get this test running on my laptop (haven't tried to build llvm on mac before), while waiting for more feedbacks. I'm thinking of submitting it on Thursday or Friday.  @ellishg I think the added compiler-rt test (on macosx) should be a test case for issue 74565.

mingmingl-llvm committed:
run git merge main to resolve merge conflict

mingmingl-llvm committed:
In compiler-rt test, make dso_local optional and the matcher stricter
for disassembled IR.
- Currently, assume dso_local is true on ELF and COFF according to https://github.com/llvm/llvm-project/blob/4f1ddf7523c0bbb4075b1682dbe2278080642eee/clang/lib/CodeGen/CodeGenModule.cpp#L1528
- Running the compiler-rt test on mac would fail on `dso_local` without this. The test should fail on imports of indirect callees as well as finding `PGONameMetadata` for external functions.

minglotus-6 commented:
> Resolved comments except the pending discussion about whether to use `.proftext` or`.profraw`.
> 
> Added `REQUIRES: zlib` for LLVM IR test since the profile reader should be built with zlib support.
> 
> I'll probably spend sometime to get this test running on my laptop (haven't tried to build llvm on mac before), while waiting for more feedbacks. I'm thinking of submitting it on Thursday or Friday. @ellishg I think the added compiler-rt test (on macosx) should be a test case for issue 74565.

The test failed on mac, but initially on trying to find `dso_local` of line `PGOName: define dso_local void @_Z7callee1v() {{.*}} !prof ![[#]] !PGOFuncName ![[#]] `.
-  Debugging with `lldb` points to [this comment](https://github.com/llvm/llvm-project/blob/4f1ddf7523c0bbb4075b1682dbe2278080642eee/clang/lib/CodeGen/CodeGenModule.cpp#L1528), saying only COFF and ELF assumes `dso_local`. 

After making the `dso_local` optional in the [latest commit](https://github.com/llvm/llvm-project/pull/74008/commits/f6a718d80d7a70d60ef8affc282f5111125e8282), the test started to fail due to the difference of `<mangled-name>` and `<linkage-name>`. 
1. Test failed, when external func `callee1` in expected output doesn't have `!PGOFuncName` metadata but the test output has it. Note `!PGOFuncName` is annotated [only if](https://github.com/llvm/llvm-project/blob/90c23981768713736208d76578ca119a20f6ac60/llvm/lib/ProfileData/InstrProf.cpp#L1262-1265) `PGOName != Func.getName()`
2. Fixing the first error allows me to see the missed import errors.

mingmingl-llvm committed:
remove a header introduced in the middle but and unused now
